File: shared/ios/ReleafCore/Sources/ReleafCoreSync/SyncDataSource.swift
/*
 * SyncDataSource.swift
 *
 * The seam that lets `SyncRepository` (in this same module, arriving in
 * PR #3b) serve both Releaf and QuickInk without knowing either app's
 * tables.
 *
 * Each app implements this protocol against its own SQLite store.
 * `SyncRepository` orchestrates a sync pass — manifest read, version
 * gate, push dirty rows, push tombstones, pull delta, write manifest —
 * and asks the data source for everything that's app-specific:
 *
 *   - "What rows are dirty?"           → nextDirtyBatch
 *   - "What deletions need pushing?"   → nextTombstoneBatch
 *   - "Apply this remote upsert."      → applyRemoteUpsert
 *   - "Apply this remote tombstone."   → applyRemoteTombstone
 *   - "These IDs uploaded clean."      → markSynced
 *
 * The contract intentionally speaks bytes, not rows. The data source is
 * responsible for decoding/encoding the canonical-JSON payloads
 * documented in docs/DRIVE_SCHEMA.md — the sync repo never sees a
 * column name and never opens a database connection. That's what lets
 * Releaf and QuickInk keep diverging schemas while sharing the
 * orchestrator.
 *
 * `kind` is a string, not an enum, for the same reason: an enum here
 * would force the shared module to know about every app's entity
 * kinds (Releaf has notebooks, chapters, pages, tasks, …; QuickInk
 * adds ocr_results). A string keeps the contract open and the
 * manifest layout uniform.
 *
 * Cursors are opaque so the data source can paginate however it likes
 * (per-table, per-updated_at, per-shard) without leaking that detail.
 *
 * See docs/QUICKINK_DESIGN.md §1 for the full design rationale + the
 * Releaf and QuickInk implementation notes.
 */

import Foundation

// MARK: - Protocol

public protocol SyncDataSource: Sendable {

    // ─── Identity / configuration ──────────────────────────────────────

    /// Top-level Drive folder name. Releaf returns `"Releaf"`, QuickInk
    /// returns `"QuickInk"`. The sync repository concatenates this with
    /// the manifest path it computes.
    var driveRootFolderName: String { get }

    /// Schema version this app currently speaks. Embedded into the
    /// manifest. A peer device that reads a manifest with a major
    /// version higher than its own returns `versionBlocked: true` from
    /// the sync pass — the user gets prompted to update the app.
    var schemaVersion: SchemaVersion { get }

    /// Stable identifier for the app — `"releaf"` or `"quickink"`.
    /// Tagged onto manifest entries so multi-app future debugging
    /// (e.g. inspecting a Drive folder shared between sibling apps)
    /// stays sane.
    var appId: String { get }

    // ─── Outbound: collect dirty rows ──────────────────────────────────

    /// Snapshot the next chunk of locally-dirty entities the sync worker
    /// should push this round. Implementations stream from each entity
    /// kind's dirty index in `updated_at` order. The worker keeps
    /// calling until it gets back a batch with `nextCursor == nil`.
    ///
    /// Pagination is data-source-defined: the cursor is opaque to the
    /// worker. A typical implementation packs `(entity_kind, updated_at,
    /// id)` so each new call resumes after the last row from the
    /// previous batch.
    ///
    /// `limit` is a hint — implementations may return fewer rows (e.g.
    /// to keep Drive payload size bounded) but should not return more.
    func nextDirtyBatch(after cursor: SyncCursor?, limit: Int) async throws -> DirtyBatch

    /// Snapshot the next chunk of tombstones to propagate. Same paging
    /// shape as `nextDirtyBatch` but the cursor namespace is independent.
    func nextTombstoneBatch(after cursor: SyncCursor?, limit: Int) async throws -> TombstoneBatch

    // ─── Inbound: apply remote changes ─────────────────────────────────

    /// Apply a remote upsert. The data source is responsible for:
    ///   1. Decoding `change.payload` per its own canonical schema
    ///      (the bytes are the same canonical JSON it emitted on the
    ///      uploading side).
    ///   2. Last-write-wins comparison against the local row by
    ///      `updatedAt` — local wins ties.
    ///   3. Clearing the local `dirty` bit if the remote write
    ///      supersedes ours.
    ///   4. Stamping `drive_file_id` from the manifest path if it's
    ///      a new local row.
    ///
    /// Idempotent: if the local row is already at this version the
    /// data source returns without writing.
    func applyRemoteUpsert(_ change: RemoteUpsert) async throws

    /// Apply a remote tombstone. Soft-delete the local row matching
    /// `(kind, id)` by setting `deleted_at = tombstone.deletedAt`,
    /// `dirty = 0`. Idempotent.
    func applyRemoteTombstone(_ tombstone: RemoteTombstone) async throws

    // ─── Bookkeeping ───────────────────────────────────────────────────

    /// Mark the given local rows as cleanly synced. Called by the worker
    /// after a successful Drive upload, in batches.
    ///
    /// For each ack: set `dirty = 0` and stamp `drive_file_id` if newly
    /// uploaded. The data source MUST do this race-safely — only flip
    /// to `dirty = 0` if `updated_at` still matches the value reported
    /// in the original `DirtyEntry` (otherwise the row was modified
    /// between the read and the upload, and we shouldn't clear the bit).
    func markSynced(_ acks: [SyncAck]) async throws

    /// Most recent remote manifest etag the app has applied. Lets the
    /// worker skip the full pull when Drive hasn't changed. Nil on
    /// first sync after install.
    func lastAppliedManifestEtag() async throws -> String?

    /// Persist the etag of the manifest the worker just applied. Called
    /// at the end of a successful sync pass.
    func setLastAppliedManifestEtag(_ etag: String) async throws
}

// MARK: - Identity types

/// Schema version embedded in the Drive manifest. Major bumps gate
/// cross-version reads; minor bumps are forward-compatible additions.
public struct SchemaVersion: Equatable, Sendable, Codable {
    public let major: Int
    public let minor: Int

    public init(major: Int, minor: Int) {
        self.major = major
        self.minor = minor
    }
}

// MARK: - Outbound types

/// A single locally-dirty entity ready to upload. The data source did
/// the canonical-JSON serialisation; the sync worker just moves bytes.
public struct DirtyEntry: Sendable, Equatable {
    /// Stable, app-defined kind tag. Examples:
    /// `"notebook"`, `"chapter"`, `"page"`, `"notepad_entry"`,
    /// `"capture"`, `"task"`, `"reminder"`, `"ocr_result"`.
    /// Used as the manifest bucket and as the routing key for inbound
    /// upserts on the receiving side.
    public let kind: String

    /// UUIDv7. Stable across devices. NOT Drive's `fileId`.
    public let id: String

    /// Path under the Drive root where this entity's bytes land.
    /// Computed by the data source per the per-kind layout in
    /// docs/DRIVE_SCHEMA.md (e.g. `notebooks/{nb}/{ch}/{page}.json`,
    /// `notepad_entries/{yyyy}/{mm}/{id}.json`).
    public let drivePath: String

    /// Canonical-JSON bytes. Sort-keyed, no extraneous whitespace —
    /// the same bytes the sha256 below was computed over.
    public let payload: Data

    /// SHA-256 of `payload`, hex-encoded. Used for manifest entries
    /// and for skip-if-unchanged comparison against the remote
    /// manifest.
    public let payloadSha256: String

    /// `updated_at` from the source row, ISO-8601 UTC with ms.
    /// Used for ordering across batches and for race-safe
    /// `markSynced`.
    public let updatedAt: String

    public init(
        kind: String,
        id: String,
        drivePath: String,
        payload: Data,
        payloadSha256: String,
        updatedAt: String
    ) {
        self.kind = kind
        self.id = id
        self.drivePath = drivePath
        self.payload = payload
        self.payloadSha256 = payloadSha256
        self.updatedAt = updatedAt
    }
}

/// One page of dirty entries from `nextDirtyBatch`.
public struct DirtyBatch: Sendable, Equatable {
    public let entries: [DirtyEntry]

    /// nil → no more pages; otherwise pass back into the next call.
    public let nextCursor: SyncCursor?

    public init(entries: [DirtyEntry], nextCursor: SyncCursor?) {
        self.entries = entries
        self.nextCursor = nextCursor
    }
}

/// A single locally-recorded deletion ready to propagate as a
/// tombstone file under `tombstones/`.
public struct TombstoneEntry: Sendable, Equatable {
    public let kind: String
    public let id: String
    /// ISO-8601 UTC timestamp when the local deletion happened.
    public let deletedAt: String

    public init(kind: String, id: String, deletedAt: String) {
        self.kind = kind
        self.id = id
        self.deletedAt = deletedAt
    }
}

public struct TombstoneBatch: Sendable, Equatable {
    public let entries: [TombstoneEntry]
    public let nextCursor: SyncCursor?

    public init(entries: [TombstoneEntry], nextCursor: SyncCursor?) {
        self.entries = entries
        self.nextCursor = nextCursor
    }
}

/// Opaque pagination cursor. The data source defines its content; the
/// sync worker treats it as a black box and round-trips it verbatim.
public struct SyncCursor: Sendable, Equatable {
    public let opaque: String

    public init(_ opaque: String) {
        self.opaque = opaque
    }
}

// MARK: - Inbound types

/// A remote upsert pulled from Drive that the data source should apply
/// to the local store. The bytes are exactly what the uploader sent —
/// canonical JSON, sort-keyed, no whitespace differences.
public struct RemoteUpsert: Sendable, Equatable {
    public let kind: String
    public let id: String
    public let payload: Data

    /// `updated_at` parsed out of the manifest entry. Used by the data
    /// source for last-write-wins.
    public let updatedAt: String

    /// Drive `fileId` of the payload file. The data source stamps this
    /// onto the local row's `drive_file_id` column when it applies.
    public let driveFileId: String

    public init(
        kind: String,
        id: String,
        payload: Data,
        updatedAt: String,
        driveFileId: String
    ) {
        self.kind = kind
        self.id = id
        self.payload = payload
        self.updatedAt = updatedAt
        self.driveFileId = driveFileId
    }
}

/// A remote tombstone pulled from Drive. The data source soft-deletes
/// the matching local row (setting `deleted_at`, clearing `dirty`).
public struct RemoteTombstone: Sendable, Equatable {
    public let kind: String
    public let id: String
    public let deletedAt: String

    public init(kind: String, id: String, deletedAt: String) {
        self.kind = kind
        self.id = id
        self.deletedAt = deletedAt
    }
}

/// One row from a successful upload. The data source flips `dirty = 0`
/// for the matching `(kind, id, updatedAt)` tuple and stamps
/// `drive_file_id`.
public struct SyncAck: Sendable, Equatable {
    public let kind: String
    public let id: String

    /// The `updatedAt` value that was on the row when it was uploaded.
    /// The data source MUST only clear the dirty bit if the local row
    /// still has this `updatedAt` — otherwise it's been edited since
    /// the upload started and we shouldn't mark it clean.
    public let updatedAt: String

    /// Drive `fileId` returned by the upload. Stamped onto the local
    /// row's `drive_file_id` column.
    public let driveFileId: String

    public init(
        kind: String,
        id: String,
        updatedAt: String,
        driveFileId: String
    ) {
        self.kind = kind
        self.id = id
        self.updatedAt = updatedAt
        self.driveFileId = driveFileId
    }
}

File: shared/android/shared/sync/src/main/kotlin/app/releaf/mobile/data/sync/SyncDataSource.kt
/*
 * SyncDataSource.kt
 *
 * The seam that lets `SyncRepository` (in this same module) serve both
 * Releaf and QuickInk without knowing either app's tables.
 *
 * Mirror of iOS's `SyncDataSource` Swift protocol — same shape, same
 * contract, same bytes-not-rows boundary. Each app implements this
 * interface against its own Room DAOs:
 *
 *   - Releaf  → ReleafSyncDataSource (apps/releaf/android/app/...)
 *   - QuickInk → QuickInkSyncDataSource (Phase 3, not yet present)
 *
 * The orchestrator hands the data source canonical-JSON bytes to apply
 * and asks for canonical-JSON bytes to upload. It never sees a column
 * name and never opens a database connection. That's what lets the two
 * apps keep diverging schemas while sharing the orchestrator.
 *
 * Why an interface, not an abstract class:
 *   - Plays nicer with Kotlin's `by` delegation if a future implementation
 *     wants to wrap another data source.
 *   - All operations are `suspend fun`s — no shared state across calls,
 *     no class invariants worth carrying.
 *
 * See docs/QUICKINK_DESIGN.md §1 for the full design rationale.
 */

@file:Suppress("unused") // public surface consumed across modules

package app.releaf.mobile.data.sync

interface SyncDataSource {

    // ─── Identity / configuration ──────────────────────────────────────

    /**
     * Top-level Drive folder name. Releaf returns `"Releaf"`, QuickInk
     * returns `"QuickInk"`. The sync repository concatenates this with
     * the manifest path it computes.
     */
    val driveRootFolderName: String

    /**
     * Schema version this app currently speaks. Embedded into the
     * manifest. A peer device that reads a manifest with a major
     * version higher than its own returns `versionBlocked = true` from
     * the sync pass — the user gets prompted to update the app.
     */
    val schemaVersion: SchemaVersion

    /**
     * Stable identifier for the app — `"releaf"` or `"quickink"`.
     * Tagged onto manifest entries so multi-app future debugging
     * (e.g. inspecting a Drive folder shared between sibling apps)
     * stays sane.
     */
    val appId: String

    // ─── Outbound: collect dirty rows ──────────────────────────────────

    /**
     * Snapshot the next chunk of locally-dirty entities the sync worker
     * should push this round. Implementations stream from each entity
     * kind's dirty index in `updated_at` order. The worker keeps
     * calling until it gets back a batch with `nextCursor == null`.
     *
     * Pagination is data-source-defined; the cursor is opaque to the
     * worker. A typical implementation packs `(entity_kind, updated_at,
     * id)` so each new call resumes after the last row of the previous
     * batch.
     *
     * `limit` is a hint — implementations may return fewer rows (e.g.
     * to keep Drive payload size bounded) but should not return more.
     */
    suspend fun nextDirtyBatch(after: SyncCursor?, limit: Int): DirtyBatch

    /**
     * Snapshot the next chunk of tombstones to propagate. Same paging
     * shape as [nextDirtyBatch] but the cursor namespace is independent.
     */
    suspend fun nextTombstoneBatch(after: SyncCursor?, limit: Int): TombstoneBatch

    // ─── Inbound: apply remote changes ─────────────────────────────────

    /**
     * Apply a remote upsert. The data source is responsible for:
     *   1. Decoding `change.payload` per its own canonical schema
     *      (the bytes are the same canonical JSON it emitted on the
     *      uploading side).
     *   2. Last-write-wins comparison against the local row by
     *      `updatedAt` — local wins ties.
     *   3. Clearing the local `dirty` bit if the remote write
     *      supersedes ours.
     *   4. Stamping `drive_file_id` from the manifest path if it's
     *      a new local row.
     *
     * Idempotent: if the local row is already at this version the
     * data source returns without writing.
     */
    suspend fun applyRemoteUpsert(change: RemoteUpsert)

    /**
     * Apply a remote tombstone. Soft-delete the local row matching
     * `(kind, id)` by setting `deleted_at = tombstone.deletedAt`,
     * `dirty = 0`. Idempotent.
     */
    suspend fun applyRemoteTombstone(tombstone: RemoteTombstone)

    // ─── Bookkeeping ───────────────────────────────────────────────────

    /**
     * Mark the given local rows as cleanly synced. Called by the worker
     * after a successful Drive upload, in batches.
     *
     * For each ack: set `dirty = 0` and stamp `drive_file_id` if newly
     * uploaded. The data source MUST do this race-safely — only flip
     * to `dirty = 0` if `updated_at` still matches the value reported
     * in the original [DirtyEntry] (otherwise the row was modified
     * between the read and the upload, and we shouldn't clear the bit).
     */
    suspend fun markSynced(acks: List<SyncAck>)

    /**
     * Most recent remote manifest etag the app has applied. Lets the
     * worker skip the full pull when Drive hasn't changed. Null on
     * first sync after install.
     */
    suspend fun lastAppliedManifestEtag(): String?

    /** Persist the etag of the manifest the worker just applied. */
    suspend fun setLastAppliedManifestEtag(etag: String)

    // ─── Drive-side cleanup ────────────────────────────────────────────

    /**
     * Identify entities in [manifest] whose parent isn't in the same
     * manifest's `entityChecksums` — i.e. orphans where the child JSON
     * file is still listed but the parent is gone. Used by
     * [SyncRepository.cleanupOrphans] to retire those entries on Drive.
     *
     * Typical example (QuickInk): an `ocr_result` whose `capture_id`
     * isn't in the manifest. Old builds soft-deleted captures without
     * cascading to children, so the next sync uploaded the parent
     * tombstone but left the child JSON in `entityChecksums` forever.
     * Every subsequent restore on any device sees the orphan and
     * either silently fails the FK insert (pre-PR-B/D) or harmlessly
     * skips it (post-PR-D). This hook lets us actually clean Drive
     * up.
     *
     * Default implementation returns an empty list — apps without
     * parent-child relationships in their synced kinds (or apps that
     * don't care about Drive-side cleanup) get no-op behaviour.
     */
    suspend fun findOrphanIds(manifest: ManifestV2): List<OrphanInfo> = emptyList()
}

/**
 * Pointer to a single orphan entity on Drive — the manifest's child
 * record whose parent isn't in the same manifest. Returned by
 * [SyncDataSource.findOrphanIds] and consumed by
 * [SyncRepository.cleanupOrphans].
 */
data class OrphanInfo(
    /** The orphan entity's id (UUIDv7). */
    val id: String,
    /** The orphan's kind — e.g. `DrivePath.KIND_OCR_RESULT`. */
    val kind: String,
    /**
     * The id of the parent that's missing from the manifest.
     * Logged for diagnostics; not strictly needed by the cleanup
     * path itself.
     */
    val missingParentId: String,
)

// ─── Outbound types ────────────────────────────────────────────────────

/**
 * A single locally-dirty entity ready to upload. The data source did
 * the canonical-JSON serialisation; the sync worker just moves bytes.
 */
data class DirtyEntry(
    /**
     * Stable, app-defined kind tag. Examples: `"notebook"`, `"chapter"`,
     * `"page"`, `"notepad_entry"`, `"capture"`, `"task"`, `"reminder"`,
     * `"ocr_result"`. Used as the manifest bucket and as the routing
     * key for inbound upserts on the receiving side.
     */
    val kind: String,

    /** UUIDv7. Stable across devices. NOT Drive's `fileId`. */
    val id: String,

    /**
     * Path under the Drive root where this entity's bytes land.
     * Computed by the data source per the per-kind layout in
     * docs/DRIVE_SCHEMA.md.
     */
    val drivePath: String,

    /**
     * Canonical-JSON bytes. Sort-keyed, no extraneous whitespace —
     * the same bytes [payloadSha256] was computed over.
     */
    val payload: ByteArray,

    /** SHA-256 of [payload], hex-encoded. */
    val payloadSha256: String,

    /** `updated_at` from the source row, ISO-8601 UTC with ms. */
    val updatedAt: String,
) {
    // Data class equality over ByteArray uses reference equality by default;
    // override for value semantics.
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is DirtyEntry) return false
        return kind == other.kind &&
            id == other.id &&
            drivePath == other.drivePath &&
            payload.contentEquals(other.payload) &&
            payloadSha256 == other.payloadSha256 &&
            updatedAt == other.updatedAt
    }

    override fun hashCode(): Int {
        var result = kind.hashCode()
        result = 31 * result + id.hashCode()
        result = 31 * result + drivePath.hashCode()
        result = 31 * result + payload.contentHashCode()
        result = 31 * result + payloadSha256.hashCode()
        result = 31 * result + updatedAt.hashCode()
        return result
    }
}

/** One page of dirty entries from [SyncDataSource.nextDirtyBatch]. */
data class DirtyBatch(
    val entries: List<DirtyEntry>,
    /** null → no more pages; otherwise pass back into the next call. */
    val nextCursor: SyncCursor?,
)

/**
 * A single locally-recorded deletion ready to propagate as a tombstone
 * file under `tombstones/`. Distinct from [TombstoneEntry] (in
 * Manifest.kt) — that one is the Drive wire format where the entity id
 * is the dictionary key. This is the data source's outbound batch
 * element, where the id is a field on the value because we're returning
 * a list, not a map.
 */
data class PendingTombstone(
    val kind: String,
    val id: String,
    /** ISO-8601 UTC timestamp when the local deletion happened. */
    val deletedAt: String,
)

data class TombstoneBatch(
    val entries: List<PendingTombstone>,
    val nextCursor: SyncCursor?,
)

/**
 * Opaque pagination cursor. The data source defines its content; the
 * sync worker treats it as a black box and round-trips it verbatim.
 */
@JvmInline
value class SyncCursor(val opaque: String)

// ─── Inbound types ─────────────────────────────────────────────────────

/**
 * A remote upsert pulled from Drive that the data source should apply
 * to the local store. The bytes are exactly what the uploader sent —
 * canonical JSON, sort-keyed, no whitespace differences.
 */
data class RemoteUpsert(
    val kind: String,
    val id: String,
    val payload: ByteArray,
    /** `updated_at` parsed out of the manifest entry. */
    val updatedAt: String,
    /** Drive `fileId` of the payload file. Stamped onto the local row. */
    val driveFileId: String,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RemoteUpsert) return false
        return kind == other.kind &&
            id == other.id &&
            payload.contentEquals(other.payload) &&
            updatedAt == other.updatedAt &&
            driveFileId == other.driveFileId
    }

    override fun hashCode(): Int {
        var result = kind.hashCode()
        result = 31 * result + id.hashCode()
        result = 31 * result + payload.contentHashCode()
        result = 31 * result + updatedAt.hashCode()
        result = 31 * result + driveFileId.hashCode()
        return result
    }
}

/**
 * A remote tombstone pulled from Drive. The data source soft-deletes
 * the matching local row.
 */
data class RemoteTombstone(
    val kind: String,
    val id: String,
    val deletedAt: String,
)

/**
 * One row from a successful upload. The data source flips `dirty = 0`
 * for the matching `(kind, id, updatedAt)` tuple and stamps
 * `drive_file_id`.
 */
data class SyncAck(
    val kind: String,
    val id: String,
    /**
     * The `updatedAt` value that was on the row when it was uploaded.
     * The data source MUST only clear the dirty bit if the local row
     * still has this `updatedAt` — otherwise it's been edited since
     * the upload started.
     */
    val updatedAt: String,
    /** Drive `fileId` returned by the upload. */
    val driveFileId: String,
)
